source/_posts/2016-12-29-testing-against-external-services.markdown

```diff
@@ -0,0 +1,91 @@
+---
+layout: post
+title: "Testing Against External Services"
+date: 2016-12-29 12:07:07 -0800
+comments: true
+author: "Lindsey Anne <lindsey.hogg@grandrounds.com>"
+categories: testing, rspec
+---
+Testing does not have to be complicated, but sometimes it can turn into spaghetti just like any other code. [This talk](https://www.youtube.com/watch?v=URSWYvyc42M) by Sandi Metz at Railsconf 2013 inspired me to revisit some of the tests for one of our services that relies heavily on FTP and SFTP connections.
+
+##### Why are my tests failing?
+
+The previous test suite for this service used a gem that was originally created as a mocked framework for testing FTP connection behaviors. However, a few key problems existed with this solution:
+
+- The gem hasn’t been maintained in a couple years, which started to cause intermittent spec failures.
+- The gem doesn't support SFTP connections, which our service relies much more heavily on than FTP.
+- The gem doesn't actually mock anything--it starts up a local FTP server, runs the tests, and then shuts the server down. Each test that’s run against it is generating a real response, albeit locally. This not only took up a lot of time (starting and stopping the local server before and after each spec), but was also another source of intermittent failures.
+
+So, after creating a plan of attack, I rewrote or deleted every single test that used this gem. Each test was also re-evaluated for "sighting along the space capsule," as well—any test that was written for a private method, a message sent to the object’s self, or an implementation detail was scrutinized (and most of them deleted), which exposed the true underlying issue: almost every single test we'd written that used the gem was an *implementation* test, not an *interface* test.
+
+External responses are supposed to be completely mocked out in testing, not make actual calls to an API or server. Tests exist to make sure OUR code works as expected, not to verify that no errors are raised when making external requests. A test failing because an external response was incorrect is a broken test already.
```

Comment:
`A test failing because an external response was incorrect is a broken test already`.  We should have a test that exercises what our code does when it gets nonsense answers.  I think what you're getting at here is something like `A test that requires a response from another service is already a broken test`